File: 프로그래머스/lv3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.cpp
#include <iostream>
#include <string>
using namespace std;
bool sliding(string& s, int len, int start){
    if(len ==1)
        return true;
    int end = start + len -1;
    int mid = start +len /2;
    while(start<=mid)
    {
        if(s[start] != s[end])
            return false;
        start ++;
        end--;
    }
    return true;
}
int solution(string s)
{
    int min = 1;
    int max = s.size();
    while(true){
        for (int len = max;len>=1;){
            for(size_t idx = 0;idx<s.length();idx++){
                if(idx+len > s.size())
                    break;
                if(sliding(s,len,idx)){
                    min = len;
                    return min;
                        
                }
            }
        len = len-1;
        }
        return min;
    }
    return min;
}

File: 프로그래머스/lv3/12904. 가장 긴 팰린드롬/가장 긴 팰린드롬.js
function solution(s)
{
    let answer = 1
    const S_LEN = s.length
    const dp = new Array(S_LEN).fill().map(_ => new Array(S_LEN).fill(false))
    
    // 길이가 1, 2
    for(let i=0; i<S_LEN; i++) {
        dp[i][i] = true
        if(i+1 < S_LEN && s[i] === s[i+1]) {
            dp[i][i+1] = true
            answer = 2
        }
    }

    // 길이 3이상
    for(let i=3; i<=S_LEN; i++) {
        for(let start=0; start<=S_LEN-i; start++) {
            const end = start+i-1
            if(s[start] === s[end] && dp[start+1][end-1]) {
                dp[start][end] = true
                answer = answer > i ? answer : i
            }
        }
    }
    
    return answer
}